src/StoreApi/Authentication.php

```diff
@@ -20,15 +22,62 @@ public function init() {
 	 * @return \WP_Error|null|bool
 	 */
 	public function check_authentication( $result ) {
-		if ( ! empty( $result ) ) {
+		if ( ! $this->is_request_to_store_api() ) {
 			return $result;
 		}
 
-		if ( $this->is_request_to_store_api() ) {
-			return true;
+		$rate_limiting_options = RateLimits::get_options();
+
+		if ( $rate_limiting_options->enabled ) {
+			$action_id = 'store_api_request_';
+
+			if ( is_user_logged_in() ) {
+				$action_id .= get_current_user_id();
+			} else {
+				$ip_address = self::get_ip_address( $rate_limiting_options->proxy_support );
+
+				if ( ! $ip_address ) {
+					return new \WP_Error(
+						'ip_address_cannot_be_determined',
+						'Bad request. Client IP address cannot be determined.',
+						array( 'status' => 400 )
+					);
```

Comment:
Just throwing ideas here, but I was thinking if we treat cases with no IP address as a group... so any request that comes from a logged-out user where the IP cannot be determined for any reason shares the same rate limiting, we could allow more requests through—that would unblock add to cart requests somewhat.

Additionally, I wonder if in preference to IP address, could we use the cart session ID/customer ID to identify users making requests?

So in order, we'd use User ID, Customer Session ID, Customer IP Address.

cc @opr any thoughts on this too?

Reply:
In my understanding, we're rejecting requests without an IP, I think it's safe to assume all requests will have an IP unless explicitly malicious.

I very much like the User ID, Session ID, IP flow, but it can be followed up with.

Reply:
> I think it's safe to assume all requests will have an IP unless explicitly malicious.

I also think this. If a request somehow doesn't have an IP, then there could be other issues further down the line, e.g. processing payment might not work, for example Stripe logs the IP of the customer when paying. I don't know how to spoof not having an IP so can't test it, but I don't imagine Stripe would allow it to go through? 

> So in order, we'd use User ID, Customer Session ID, Customer IP Address.

I like this idea!

Reply:
Can't someone just delete cookies and reset the session with each request?
We already use User ID if available

Reply:
good point, a session id wouldn't catch multiple attacks, it would just slightly change an attacker's flow but wouldn't block him.

Reply:
Marking this as resolved then, User Id -> IP Address is our order preference for tracking rate limits 🙂